modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java

```diff
@@ -282,6 +282,29 @@ private MapStream(Stream<Map.Entry<K, V>> underlying) {
     return underlying.map(e -> mapper.apply(e.getKey(), e.getValue()));
   }
 
+  //-------------------------------------------------------------------------
+  /**
+   * Transforms the entries in the stream by applying a mapper function to each key.
+   *
+   * @param mapper  a mapper function whose return values are included in the new stream
+   * @param <R>  the type of elements in the new stream
+   * @return a stream containing the values returned from the mapper function
+   */
+  public <R> Stream<R> flatMapKeys(Function<? super K, Stream<? extends R>> mapper) {
+    return underlying.flatMap(e -> mapper.apply(e.getKey()));
```

Comment:
I was expecting this to return a `MapStream`. This could just be `keys().flatMap()`, so doesn't gain enough.

Reply:
So you'd be expecting something similar to this?
```
  public <nK, nV> MapStream<nK, nV> flatMapKeys(Function<? super K, Stream<Pair<nK, nV>>> mapper) {
    return wrap(underlying
        .flatMap(e -> mapper.apply(e.getKey()))
        .map(pair -> entry(pair.getFirst(), pair.getSecond())));
  }
```

Or if Pair implemented Map.Entry then can remove the last .map call and somebody would be able to pass in a MapStream into the function

i.e
```
  public <nK, nV> MapStream<nK, nV> flatMapKeys(Function<? super K, Stream<Map.Entry<nK, nV>>> mapper) {
    return wrap(underlying.flatMap(e -> mapper.apply(e.getKey()));
  }

MapStream.of(map)
        .filterValues(this::someValueFilter)
        .flatMapKeys(key -> MapStream.of(nestedMaps.get(key)))
```

Reply:
I'd expect `flatMapKeys` to take each key, flat map it, and associate the original value with each new key. `MapStream<List<String>, Integer>` ->  `MapStream<String, Integer>`.

I'm not sure that `flatMapValues` is feasible.

Reply:
@wjnicholson what's the use case that inspired these?

Reply:
The original reasoning was flattenening nested data structures (that contained nested maps), e.g
` MapStream<Key, Map<Key2, Map<Key3, Value>>> ` -> ` Map<Key, Value>`

Reply:
I also had in the back of my head a want for MapStream like semantics for a PairStream, in particular the ` Guavate.zip() ` and ` Guavate.zipWithIndex() ` methods. Since we currently have no way of transforming a Stream<Pair> into a MapStream.

Reply:
True, I guess I'm being pernickety about the verbosity with method references.